rust/lance-core/src/io/writer/statistics.rs

```diff
@@ -22,102 +22,421 @@ use crate::datatypes::Field;
+        if v.len() > BINARY_PREFIX_LENGTH {
+            max_value = truncate_utf8(v, BINARY_PREFIX_LENGTH);
+            max_value_bound = increment_utf8(max_value.unwrap().as_bytes().to_vec()).unwrap();
+            max_value = Some(str::from_utf8(&max_value_bound).unwrap());
+        }
+    }
+
+    let min_value = min_value.map(|x| x.to_string());
+    let max_value = max_value.map(|x| x.to_string());
+
+    match arrays[0].data_type() {
+        DataType::Utf8 => StatisticsRow {
+            null_count: ScalarValue::Int64(Some(null_count)),
+            min_value: ScalarValue::Utf8(min_value),
+            max_value: ScalarValue::Utf8(max_value),
+        },
+        DataType::LargeUtf8 => StatisticsRow {
+            null_count: ScalarValue::Int64(Some(null_count)),
+            min_value: ScalarValue::LargeUtf8(min_value),
+            max_value: ScalarValue::LargeUtf8(max_value),
+        },
+        _ => {
+            todo!()
+        }
+    }
+}
+
+fn get_binary_statistics<T: OffsetSizeTrait>(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value: Option<&[u8]> = None;
+    let mut max_value: Option<&[u8]> = None;
+    let mut null_count: i64 = 0;
+
+    let array_iterator = arrays.iter().map(|x| as_generic_binary_array::<T>(x));
+
+    for array in array_iterator {
+        null_count += array.null_count() as i64;
+        if array.null_count() == array.len() {
+            continue;
+        }
+
+        array.iter().for_each(|value| {
+            if let Some(val) = value {
+                // don't compare full buffers if possible
+                let val = &val[..std::cmp::min(BINARY_PREFIX_LENGTH + 4, val.len())];
+
+                if let Some(v) = min_value {
+                    if let Some(Ordering::Less) = val.partial_cmp(v) {
+                        min_value = Some(val);
+                    }
+                } else {
+                    min_value = Some(val);
+                }
+
+                if let Some(v) = max_value {
+                    if let Some(Ordering::Greater) = val.partial_cmp(v) {
+                        max_value = Some(val);
+                    }
+                } else {
+                    max_value = Some(val);
+                }
+            }
+        });
+    }
+
+    if let Some(v) = min_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            min_value = truncate_binary(v, BINARY_PREFIX_LENGTH);
+        }
+    }
+
+    let max_value_bound: Vec<u8>;
+    if let Some(v) = max_value {
+        if v.len() > BINARY_PREFIX_LENGTH {
+            max_value = truncate_binary(v, BINARY_PREFIX_LENGTH);
+            if let Some(x) = increment(max_value.unwrap().to_vec()) {
+                max_value_bound = x;
+                max_value = Some(&max_value_bound);
+            } else {
+                max_value = None;
+            }
+        }
+    }
+
+    let min_value = min_value.map(|x| x.to_vec());
+    let max_value = max_value.map(|x| x.to_vec());
```

Comment:
Right now you are allocated the max twice if truncating.
```suggestion
    let min_value = min_value.map(|x| x.to_vec());
    let max_value = if let Some(v) = max_value {
        if v.len() > BINARY_PREFIX_LENGTH {
            max_value = truncate_binary(v, BINARY_PREFIX_LENGTH);
            if let Some(x) = increment(max_value.unwrap().to_vec()) {
                Some(x)
            } else {
                None
            }
        } else {
            Some(v.to_vec())
        }
    } else {
        None
    }    
```